src/main/java/net/snowflake/ingest/streaming/internal/ParquetFlusher.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2022 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.streaming.internal;
+
+import static net.snowflake.ingest.utils.Constants.MAX_CHUNK_SIZE_IN_BYTES;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.Logging;
+import net.snowflake.ingest.utils.SFException;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.hadoop.ParquetFileWriter;
+import org.apache.parquet.hadoop.ParquetWriter;
+import org.apache.parquet.hadoop.api.WriteSupport;
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+import org.apache.parquet.io.DelegatingPositionOutputStream;
+import org.apache.parquet.io.OutputFile;
+import org.apache.parquet.io.ParquetEncodingException;
+import org.apache.parquet.io.PositionOutputStream;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.io.api.RecordConsumer;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.PrimitiveType;
+
+/**
+ * Converts {@link ChannelData} buffered in {@link RowBuffer} to the Parquet format for faster
+ * processing.
+ */
+public class ParquetFlusher implements Flusher<ParquetRowBuffer.ParquetChunkData> {
+  private static final Logging logger = new Logging(ParquetFlusher.class);
+  private final MessageType schema;
+
+  public ParquetFlusher(MessageType schema) {
+    this.schema = schema;
+  }
+
+  @Override
+  public SerializationResult serialize(
+      List<ChannelData<ParquetRowBuffer.ParquetChunkData>> channelsDataPerTable,
+      ByteArrayOutputStream chunkData,
+      String filePath)
+      throws IOException {
+    List<ChannelMetadata> channelsMetadataList = new ArrayList<>();
+    long rowCount = 0L;
+    SnowflakeStreamingIngestChannelInternal<ParquetRowBuffer.ParquetChunkData> firstChannel = null;
+    Map<String, RowBufferStats> columnEpStatsMapCombined = null;
+    List<List<Object>> rows = null;
+
+    for (ChannelData<ParquetRowBuffer.ParquetChunkData> data : channelsDataPerTable) {
+      // Create channel metadata
```

Comment:
Question about everything in the `for` loop here: a lot (all?) of it looks to be non-Parquet specific. Is this the same in the Arrow flusher? If so, would it be possible to refactor it to make the amount of duplication smaller?

Reply:
I thought about it but this does not look very straight forward for me atm.
There are common pieces but scattered around, I would get back this if we see it growing.